qlty-check/src/planner/plugin_workspace_entry_finder_builder.rs

```diff
@@ -91,15 +94,33 @@ impl PluginWorkspaceEntryFinderBuilder {
             )));
         }
 
-        let ignores_without_metadata = self
-            .ignores
+        let exclude_groups = ExcludeGroup::build_from_excludes(&self.excludes.iter().collect());
+
+        matchers.push(Box::new(ExcludeGroupsMatcher::new(exclude_groups)));
+
+        let all_excluded_plugins: HashSet<String> = self
+            .excludes
             .iter()
-            .filter(|i| i.plugins.is_empty() && i.rules.is_empty() && i.levels.is_empty())
+            .flat_map(|i| i.plugins.clone())
             .collect();
 
-        let ignore_groups = IgnoreGroup::build_from_ignores(&ignores_without_metadata);
+        for plugin_name in all_excluded_plugins {
+            let plugin_specific_excludes = self
+                .excludes
+                .iter()
+                .filter(|i| i.plugins.contains(&plugin_name))
+                .cloned()
+                .collect::<Vec<_>>();
+
+            if !plugin_specific_excludes.is_empty() {
+                matchers.push(Box::new(PluginSpecificExcludeMatcher::new(
+                    plugin_name.clone(),
+                    plugin_specific_excludes,
+                    self.root.clone(),
+                )));
+            }
+        }
```

Comment:
This section of code reads a bit unexpected and somewhat confusing to me.

Is this some form of optimization?